src/host.cpp

```diff
@@ -644,21 +614,21 @@ namespace rhost {
                     }
                 }
 
-                if (!msg.request_id.empty()) {
-                    if (msg.request_id != id) {
-                        fatal_error("Received response ['%s','%s'], while awaiting response for ['%s','%s'].",
-                            msg.request_id.c_str(), msg.name.c_str(), id.c_str(), name);
-                    } else if (strcmp(msg.name.c_str() + 1, name + 1) != 0) {
-                        fatal_error("Response to ['%s','%s'] has mismatched name '%s'.",
-                            id.c_str(), name, msg.name.c_str());
+                if (msg.is_response()) {
+                    if (msg.request_id() != id) {
+                        fatal_error("Received response [%llu,'%s'], while awaiting response for [%llu,'%s'].",
+                            msg.request_id(), msg.name(), id, name.c_str());
+                    } else if (strcmp(msg.name() + 1, name.c_str() + 1) != 0) {
+                        fatal_error("Response to [%llu,'%s'] has mismatched name '%s'.",
+                            id, name.c_str(), msg.name());
                     }
                     return msg;
                 }
 
-                if (msg.name.size() >= 2 && msg.name[0] == '?' && msg.name[1] == '=') {
+                if (msg.name()[0] == '?' && msg.name()[1] == '=') {
```

Comment:
`name()` now returns `const char*`, and it's a null-terminated string. So this is guaranteed to work due to short circuiting behavior - there's always [0] to compare, and if the string is 0-length, it'll be `\0`, so first comparison fails, and second never happens. If first comparison succeeds, then [1] is guaranteed to be there (but is possibly `\0`). Etc.